File: utils/benchmark/BST/bst.py
#!/usr/bin/env python
#
# A binary search tree.
class BST:
    class Node:
        left, right, key = None, None, 0
    
        def __init__(self, key):
            self.left = None
            self.right = None
            self.key = key

    def __init__(self):
        self.root = None
    
    def insert(self, key):
        if not self.root:
            self.root = BST.Node(key)
            return True
        return self.insertInto(key, self.root)

    def insertInto(self, key, node):
        if key == node.key:
            return False
        if key < node.key:
            if node.left:
                return self.insertInto(key, node.left)
            else:
                node.left = BST.Node(key)
                return True
        else:
            if node.right:
                return self.insertInto(key, node.right)
            else:
                node.right = BST.Node(key)
                return True

    def find(self, key):
        return self.findIn(key, self.root)

    def findIn(self, key, node):
        if not node:
            return False
        if key == node.key:
            return True
        if key < node.key:
            return self.findIn(node.left, key)
        else:
            return self.findIn(node.right, key)

    def minKey(self):
        if not self.root:
            return None
        node = self.root
        while (node.left):
            node = node.left
        return node.key

    def maxKey(self):
        if not self.root:
            return None
        node = self.root
        while (node.right):
            node = node.right
        return node.key

    def depth(self):
        return self.depthIn(self.root)
        
    def depthIn(self, node):
        if not node:
            return 0
        leftDepth = self.depthIn(node.left)
        rightDepth = self.depthIn(node.right)
        return max(leftDepth, rightDepth) + 1
            
    def size(self):
        return self.sizeIn(self.root)
    
    def sizeIn(self, node):
        if not node:
            return 0
        return 1 + self.sizeIn(node.left) + self.sizeIn(node.right)

    def visitInorder(self, f):
        self.visitInorderSubtree(f, self.root)

    def visitInorderSubtree(self, f, node):
        if not node:
            return
        self.visitInorderSubtree(f, node.left)
        f(node.key)
        self.visitInorderSubtree(f, node.right)

def printKey(key):
    print key,

def usage():
    print "Usage: bst <numkeys>"

if __name__ == "__main__":
    import sys, lfsr
    if len(sys.argv) != 2:
        usage()
        sys.exit(1)

    bst = BST()
    lfsr = lfsr.LFSR()
    for i in range(0,int(sys.argv[1])):
        bst.insert(lfsr.randInt())
    
    print "Size:", bst.size()
    print "Depth:", bst.depth()
    print "Range:", bst.minKey(), "-", bst.maxKey()
    print "Values:",
    bst.visitInorder(printKey)

File: utils/benchmark/BST/bst.swift
// A binary search tree.
//
// TODO: make it generic once we handle the basics.
class BST {
  class Node {
    var left:Node?
    var right:Node?
    var key:Int = 0
    
    init(_ key:Int) {
      self.key = key
    }
  }

  var root:Node?
    
  func insert(key:Int) -> Bool {
    if !root {
      root = BST.Node(key)
      return true
    }
    return insertInto(key, node: root!)
  }

  // Written this way for tail recursion.
  func insertInto(key:Int, node:Node) -> Bool {
    if key == node.key {
      return false
    }
    if key < node.key {
      if node.left {
        return insertInto(key, node: node.left!)
      } else {
        node.left = BST.Node(key)
        return true
      }
    } else {
      if node.right {
        return insertInto(key, node: node.right!)
      } else {
        node.right = BST.Node(key)
        return true
      }
    }
  }

  func find(key:Int) -> Bool {
    return findIn(key, node: root)
  }

  func findIn(key:Int, node:Node?) -> Bool {
    if !node {
      return false
    }
    let n = node!
    if key == n.key {
      return true
    }
    if key < n.key {
      return self.findIn(key, node: n.left)
    } else {
      return self.findIn(key, node: n.right)
    }
  }

  func minKey() -> Int? {
    if !root {
      return nil
    }
    var node = root!
    while (node.left) {
      node = node.left!
    }
    return node.key
  }

  func maxKey() -> Int? {
    if !root {
      return nil
    }
    var node = root!
    while (node.right) {
      node = node.right!
    }
    return node.key
  }

  func depth() -> Int {
    return depthIn(root)
  }

  func depthIn(node:Node?) ->Int {
    if !node {
      return 0
    }
    let n = node!
    let leftDepth = depthIn(n.left)
    let rightDepth = depthIn(n.right)
    return max(leftDepth, rightDepth) + 1
  }        

  func size() -> Int {
    return sizeIn(root)
  }

  func sizeIn(node:Node?) -> Int {
    if let n = node {
      return 1 + sizeIn(n.left) + sizeIn(n.right)
    }
    return 0
  }

  func visitInorder(f:(Int) -> ()) {
    self.visitInorderSubtree(f, node: root)
  }
  func visitInorderSubtree(f:(Int) -> (), node:Node?) {
    if let n = node {
      self.visitInorderSubtree(f, node: n.left)
      f(n.key)
      self.visitInorderSubtree(f, node: n.right)
    }
  }
}

import liblfsr

func test(N:Int) {
  var bst = BST()
  var lfsr = LFSR()
  for i in 0..N {
    bst.insert(lfsr.randInt())
  }
  print("Size \(bst.size())")
  print("Depth \(bst.depth())")
  if let min = bst.minKey() {
    let max = bst.maxKey()!
    print("Range \(min) - \(max)")
  }
  print("Values: ")
  bst.visitInorder({print("\($0) ")})
  print("")
}

// I'm embarassed that I don't know how to read cmd line args.
test(1000)
